packages/core/src/types.ts

```diff
@@ -189,32 +189,73 @@ export interface Preflight<Theme extends {} = {}> {
 
 export type BlocklistRule = string | RegExp
 
+export interface VariantHandlerContext {
+  /**
+   * Rewrite the output selector. Often be used to append pesudo classes or parents.
+   */
+  selector: string
+  /**
+   * Rewrite the output css body. The input come in [key,value][] pairs.
+   */
+  entries: CSSEntries
+  /**
+   * Provide a parent selector(e.g. media query) to the output css.
+   */
+  parent?: string
+  /**
+   * Provide order to the `parent` parent selector within layer.
+   */
+  parentOrder?: number
+  /**
+   * Override layer to the output css.
+   */
+  layer?: string
+  /**
+   * Order in which the variant is sorted within single rule.
+   */
+  sort?: number
+}
+
 export interface VariantHandler {
+  /**
+   * Callback to process the handler.
+   */
+  handler?: (input: VariantHandlerContext, next: (input: VariantHandlerContext) => VariantHandlerContext) => VariantHandlerContext
   /**
    * The result rewritten selector for the next round of matching
    */
   matcher: string
   /**
+   * Order in which the variant is applied to selector.
+   */
+  order?: number
+  /**
+   * @deprecated use `handler` instead. It will be removed in 1.0.
+   *
    * Rewrite the output selector. Often be used to append pesudo classes or parents.
    */
   selector?: (input: string, body: CSSEntries) => string | undefined
   /**
+   * @deprecated use `handler` instead. It will be removed in 1.0.
+   *
    * Rewrite the output css body. The input come in [key,value][] pairs.
    */
   body?: (body: CSSEntries) => CSSEntries | undefined
   /**
+   * @deprecated use `handler` instead. It will be removed in 1.0.
+   *
    * Provide a parent selector(e.g. media query) to the output css.
    */
   parent?: string | [string, number] | undefined
   /**
-   * Order in which the variant is applied to selector.
-   */
-  order?: number
-  /**
+   * @deprecated use `handler` instead. It will be removed in 1.0.
+   *
    * Order in which the variant is sorted within single rule.
    */
   sort?: number
   /**
+   * @deprecated use `handler` instead. It will be removed in 1.0.
```

Comment:
I don't feel it's worth it to deprecate them. If we really need to be strict, we could have `type VariantHandler = VariantHandlerObject | VariantHandlerFunction` to say you can't use the object style when `handle` is provided.